### PR TITLE
Avatar - add name with initials and backgroundColor specific to that name

### PR DESCRIPTION
**Pull request by M-i-k-e-l**
## Description
Avatar - add name with initials and backgroundColor specific to that name (i.e. hash)

## Changelog
Avatar - add name with initials and backgroundColor specific to that name (i.e. hash)

### CODE REVIEW

File: src/components/avatar/index.tsx [hunk 1]
```diff
@@ -93,6 +124,20 @@ export type AvatarProps = Pick<AccessibilityProps, 'accessibilityLabel'> & {
...
+  /**
+   * The name of the avatar user.
+   * Text initials will be generated from the name.
+   * Has lower priority than label.
+   */
+  name?: string;
+  /**
+   * Override the logic for creating initials of the user name (default from AvatarHelper)
+   */
+  getInitials?: (name?: string, limit?: number) => string;
```
Comment (ethanshar) [inline src/components/avatar/index.tsx:136]: Avoid adding props that the user can have on their end. 
If the user wants to implement a custom logic for initials, they can do it and pass the `label` prop.

File: src/components/avatar/index.tsx [hunk 2]
```diff
@@ -93,6 +124,20 @@ export type AvatarProps = Pick<AccessibilityProps, 'accessibilityLabel'> & {
...
+  /**
+   * Send this to use the name to infer a backgroundColor
+   */
+  nameToColorProps?: NameToColorProps;
```
Comment (ethanshar) [inline src/components/avatar/index.tsx:140]: Maybe a better name can be colorHashConfig? autoColorsConfig? dyanmicColorsConfog? (:

File: src/components/avatar/index.tsx [hunk 3]
```diff
@@ -44,6 +46,35 @@ export enum BadgePosition {
 
 const DEFAULT_BADGE_SIZE = 'pimpleBig';
 
+export type NameToColorProps = {
+  /**
+   * Hash the name (or label) to get a color, so each name will have a specific color.
+   * Default is false.
+   */
+  useHashBackgroundColor: boolean;
```
Comment (ethanshar) [inline src/components/avatar/index.tsx:54]: missing `?`
```suggestion
  useHashBackgroundColor?: boolean;
```

Also, I think the flag prop should not be part of the config but alongside it. 

```
<Avatar useHashBackgroundColor hashColorConfig={....} name={...} />
```

File: src/components/avatar/index.tsx [hunk 4]
```diff
@@ -44,6 +46,35 @@ export enum BadgePosition {
...
+   */
+  hashFunction?: (name?: string) => number;
+  /**
+   * Background color in cases where the getBackgroundColor returns undefined.
+   */
+  defaultColor?: string;
+  /**
+   * Replace the default logic.
+   */
+  getBackgroundColor?: (
```
Comment (ethanshar) [inline src/components/avatar/index.tsx:70]: same as `getInitials`, if the user wants to have a custom logic for the background color they can implement on their end and pass thw `backgroundColor` prop

File: src/components/avatar/index.tsx [hunk 5]
```diff
@@ -44,6 +46,35 @@ export enum BadgePosition {
...
+   */
+  avatarColors?: string[];
+  /**
+   * Replace the default hashing function (name -> number)
+   */
+  hashFunction?: (name?: string) => number;
+  /**
+   * Background color in cases where the getBackgroundColor returns undefined.
+   */
+  defaultColor?: string;
```
Comment (ethanshar) [inline src/components/avatar/index.tsx:66]: See my comment on `getBackgroundColor`, not sure it's still relevant

File: src/components/avatar/index.tsx [hunk 6]
```diff
@@ -318,14 +362,61 @@ class Avatar extends PureComponent<AvatarProps> {
...
+    return text;
+  });
+
+  get text() {
+    const {label, name, getInitials = AvatarHelper.getInitials} = this.props;
+    return this.getText(label, name, getInitials);
+  }
+
+  getBackgroundColor = memoize((text,
+    useHashBackgroundColor,
```
Comment (ethanshar) [inline src/components/avatar/index.tsx:380]: Why go into this method if the user pass `useHashBackgroundColor: false`?

File: src/helpers/AvatarHelper.ts [hunk 7]
```diff
@@ -27,13 +27,13 @@ export function getColorById(id: string, avatarColors = getAvatarColors()) {
   return avatarColors[colorIndex];
 }
 
-export function getInitials(name: string) {
+export function getInitials(name?: string, limit = 2) {
```
Comment (ethanshar) [inline src/helpers/AvatarHelper.ts:30]: Consider adding proper unit test for the new limit argument.

File: src/helpers/AvatarHelper.ts [hunk 8]
```diff
@@ -43,6 +43,19 @@ export function getInitials(name: string) {
   return _.toUpper(initials);
 }
 
+export function getBackgroundColor(name?: string,
```
Comment (ethanshar) [inline src/helpers/AvatarHelper.ts:46]: consider adding few simple tests to cover the new functionalities

File: src/components/avatar/index.tsx (hunk 1 above)
Comment (M-i-k-e-l) [reply src/components/avatar/index.tsx:136]: But we want the coloring logic to only apply when `name` is sent, and `name` is ignored when `label` is sent.

File: src/components/avatar/index.tsx (hunk 2 above)
Comment (M-i-k-e-l) [reply src/components/avatar/index.tsx:140]: `autoColorsConfig` with `AutoColorsProps` 10x :)

File: src/components/avatar/index.tsx (hunk 3 above)
Comment (M-i-k-e-l) [reply src/components/avatar/index.tsx:54]: Moved (renamed to `useAutoColorsConfig`)

File: src/components/avatar/index.tsx (hunk 1 above)
Comment (M-i-k-e-l) [reply src/components/avatar/index.tsx]: Removed

File: src/components/avatar/index.tsx (hunk 4 above)
Comment (M-i-k-e-l) [reply src/components/avatar/index.tsx]: Removed

File: src/components/avatar/index.tsx (hunk 5 above)
Comment (M-i-k-e-l) [reply src/components/avatar/index.tsx:61]: I think we need this as a backup (errors etc)

File: src/components/avatar/index.tsx (hunk 6 above)
Comment (M-i-k-e-l) [reply src/components/avatar/index.tsx]: Fixed

File: src/helpers/AvatarHelper.ts (hunk 7 above)
Comment (M-i-k-e-l) [reply src/helpers/AvatarHelper.ts:30]: Added

File: src/helpers/AvatarHelper.ts (hunk 8 above)
Comment (M-i-k-e-l) [reply src/helpers/AvatarHelper.ts:46]: Done (found a bug!)

File: generatedTypes/components/avatar/index.d.ts [hunk 9]
```diff
@@ -63,6 +77,21 @@ export declare type AvatarProps = Pick<AccessibilityProps, 'accessibilityLabel'>
...
+     * The name of the avatar user.
+     * If no label is provided, the initials will be generated from the name.
+     * autoColorsConfig will use the name to create the background color of the Avatar.
+     */
+    name?: string;
+    /**
+     * Hash the name (or label) to get a color, so each name will have a specific color.
+     * Default is false.
+     */
+    useAutoColorsConfig?: boolean;
```
Comment (ethanshar) [inline generatedTypes/components/avatar/index.d.ts:90]: `useAutoColorsConfig` sounds like the user must pass it. I think `useAutoColors` is better
```suggestion
    useAutoColors?: boolean;
```

File: src/components/avatar/index.tsx [hunk 10]
```diff
@@ -318,14 +349,48 @@ class Avatar extends PureComponent<AvatarProps> {
...
+  get backgroundColor() {
+    const {backgroundColor, useAutoColorsConfig, autoColorsConfig, name} = this.props;
+    if (backgroundColor) {
+      return backgroundColor;
+    }
+
+    const {
+      avatarColors = AvatarHelper.getAvatarColors(),
+      hashFunction = AvatarHelper.hashStringToNumber,
+      defaultColor = Colors.dark80
```
Comment (ethanshar) [inline src/components/avatar/index.tsx:379]: dark preset should be deprecated eventually...
```suggestion
      defaultColor = Colors.grey80
```

File: src/helpers/AvatarHelper.ts (hunk 8 above)
Comment (ethanshar) [reply src/helpers/AvatarHelper.ts:46]: The power of testings (:

**Comment by ethanshar** [comment]
Looks great!
Last small fixes (:

File: generatedTypes/components/avatar/index.d.ts (hunk 9 above)
Comment (M-i-k-e-l) [reply generatedTypes/components/avatar/index.d.ts:90]: Done

**Review by ethanshar — APPROVED** [review]

**Pull request merged**